Re-throw template errors and buffer templates.

t3hc13h opened the pull request:
Added rethrowing of template errors in prod mode so the application can
return it's 500 code or otherwise handle the exception from the main
exception handling code. This requires full buffering of the template
output since we can't guarantee that we haven't flushed some part of the
response to the user. I think in most cases this will not matter, as the
performance tests I ran showed slightly higher GC pressure but a bit
more reqs/s with full buffering.

t3hc13h committed:
Re-throw template errors and buffer templates.

Added rethrowing of template errors in prod mode so the application can
return it's 500 code or otherwise handle the exception from the main
exception handling code. This requires full buffering of the template
output since we can't guarantee that we haven't flushed some part of the
response to the user. I think in most cases this will not matter, as the
performance tests I ran showed slightly higher GC pressure but a bit
more reqs/s with full buffering.

buildhive commented:
[Ninja Web Framework » ninja #447](https://buildhive.cloudbees.com/job/ninjaframework/job/ninja/447/) FAILURE
Looks like there's a problem with this pull request
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

t3hc13h commented:
Here's a quick summary of my performance test results using ninja-servlet-intergration-test as my benchmark project with warmed up VM (1M requests before test) on a 6 core win 8 box over 100K reqs with the results averaged over 3 runs.

Full buffering:
Collections avg: 48
Req/s avg: 1642

Direct write:
Collections avg: 39
Req/s avg: 1587

raphaelbauer commented:
That's awesome... We wanted to do some measurements here as well.. So give us some days to review and merge the PR... Needless to say that it looks really good - and it's especially cool that you measured performance before / after the change... Really good work...

raphaelbauer commented:
Hey @t3hc13h - how did you setup the benchmarking exactly?

raphaelbauer commented on ninja-core/src/main/java/ninja/template/TemplateEngineFreemarkerExceptionHandler.java:
```diff
@@ -53,38 +52,9 @@ public TemplateEngineFreemarkerExceptionHandler(Logger logger,
-
-        } else {
-            
+        if (!ninjaProperties.isProd()) {
```
I think this pw.println error will be never shown in dev or test ... The te will always be thrown, so the exception will always bubble up leading to a general error page... or did I miss something?

t3hc13h replied:
Woops! You're correct, the generic error page eats the template error page when we always re-throw.

raphaelbauer commented on ninja-core/src/main/java/ninja/template/TemplateEngineFreemarker.java:
```diff
@@ -316,12 +317,24 @@ public void invoke(Context context, Result result) {
+            // we can't guarantee we haven't flushed part of the response to the
+            // client.
+            
+            StringWriter buffer = new StringWriter(64 * 1024);
```
We need a test to make sure error handling works as we expect it to work ...

t3hc13h replied:
Agreed, I'll go ahead and make sure these changes are covered.

raphaelbauer commented on ninja-core/src/main/java/ninja/template/TemplateEngineFreemarkerExceptionHandler.java:
```diff
@@ -53,38 +52,9 @@ public TemplateEngineFreemarkerExceptionHandler(Logger logger,
 
     public void handleTemplateException(TemplateException te,
```
Not sure - but would it make sense to replace this templateerror handler with something more generic? If we have a mechanism in place that catches exceptions bubbling up - why should it not display a nicely formatted error there - with the added benefit that we can see the all kinds of exeptions in the browser while developing (not just templating expceptions)... just an idea...

t3hc13h replied:
So like a dev version of the general error page which does a message+stack print? That would be pretty cool, I'll play around with that idea and open another PR focus on that idea.

raphaelbauer commented on ninja-core/src/main/java/ninja/template/TemplateEngineFreemarker.java:
```diff
@@ -316,12 +317,24 @@ public void invoke(Context context, Result result) {
+            
+            StringWriter buffer = new StringWriter(64 * 1024);
+            
+            freemarkerTemplate.process(map, buffer);
```
Could you remove the blank lines between the statements?

t3hc13h replied:
Absolutely, while working with the class it seemed like most of the file had extra spacing, so I used spacing for consistency within the class.

raphaelbauer commented:
I adde some remarks. I hope they make sense... Let me know what you think :) I can chime in and do some stuff if you want... It's really cool polishing those parts :)

t3hc13h committed:
Fixed dev mode template error

Template exceptions are now re-thrown ONLY in production mode. Added
test coverage.

buildhive commented:
[Ninja Web Framework » ninja #458](https://buildhive.cloudbees.com/job/ninjaframework/job/ninja/458/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

buildhive commented:
[Ninja Web Framework » ninja #459](https://buildhive.cloudbees.com/job/ninjaframework/job/ninja/459/) SUCCESS
This pull request looks good
[(what's this?)](https://www.cloudbees.com/what-is-buildhive)

t3hc13h commented:
For the benchmark I used by dev machine (6 core AMD, 8GB RAM, SSD OS drive) as the server, compiled the ninja-servlet-integration-test to a single assembly and ran it directly with stdout piped to null. No particular GC or JVM memory settings, just the defaults. Client lower powered win 8 box on the local 1GB LAN.

My goal for the tests was to create a simple environment in which I could reduce the number of variables impacting the tests and measure change between the two output methods.

Before I started the warm up traffic I attached VisualVM to the process to record GC data and monitor JIT activity. For the warm up traffic I used Apache Bench  to run 1 million requests against the home route. That many requests may have been a bit over kill but that seemed to be the point the JIT activity mostly subsided so I figured that was a safe start point.

After the warm up I executed three test runs with each output method with 100K requests each to the home route three times to get my averages.

I'll admit that my tests are rather ad-hoc and I don't have much experience bench marking JVM applications so if you have a preferred method to load test Ninja or gather stats just let me know and I'll take another pass at this.

raphaelbauer commented:
Some final remarks: Could you please add one line to changelog.md? Once that's there I'll merge. Really nice improvement!

raphaelbauer commented:
forget my last comment.. I'll add the line and merge... Thanks!
